lib/datadog/di/code_tracker.rb

```diff
@@ -0,0 +1,147 @@
+# frozen_string_literal: true
+
+module Datadog
+  module DI
+    # Tracks loaded Ruby code by source file and maintains a map from
+    # source file to the loaded code (instruction sequences).
+    # Also arranges for code in the loaded files to be instrumented by
+    # line probes that have already been received by the library.
+    #
+    # The loaded code is used to target line trace points when installing
+    # line probes which dramatically improves efficiency of line trace points.
+    #
+    # Note that, since most files will only be loaded one time (via the
+    # "require" mechanism), the code tracker needs to be global and not be
+    # recreated when the DI component is created.
+    #
+    # @api private
+    class CodeTracker
```

Comment:
Might be worth documenting what parts of this class are concurrent and why.

e.g.If I understand it well (and I'm speculating/reverse engineering), the concurrency in `registry` is because `iseqs_for_path` may be concurrent with the tracepoint, but usually there will be no concurrency between multiple invocations for `iseqs_for_path` (presumably because remote configuration apply is sequential).

(I don't quite understand when there can be concurrency in `start`/`stop`; given that usually the dd-trace-rb components system takes care of enforcing that starting and stopping things is a sequential operation as well)

Reply:
start & stop should not be running at the same time. start/stop could (I suppose) be running while the trace point is invoked. Since starting and stopping should happen once and never in normal usage, I reused the mutexes.

Reply:
I'll admit that if you confirm that start/stop should not be concurrent then I don't quite understand what `trace_point_lock` is protecting :thinking: 

I saw you setup a meeting to discuss this, so let's continue the discussion there. In any case, I don't think the discussion should be a blocker to merging this PR, as we can always tweak this in a later PR.